Create IBM Cloudant resource and data source

eiri opened the pull request:
This PR adds new resource and data source for managing IBM Cloudant instances

<!--- See what makes a good Pull Request at : https://github.com/IBM-Cloud/terraform-provider-ibm/blob/master/.github/CONTRIBUTING.md --->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

<!--- If your PR fully resolves and should automatically close the linked issue, use Closes. Otherwise, use Relates --->
Relates OR Closes #0000

Output from acceptance testing:

<!--
Replace TestAccXXX with a pattern that matches the tests affected by this PR.

For more information on the `-run` flag, see the `go test` documentation at https://tip.golang.org/cmd/go/#hdr-Testing_flags.
-->
```
$ make testacc TEST=./ibm TESTARGS='-run TestAccIBMCloudant'

==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./ibm -v -run TestAccIBMCloudant -timeout 700m

[WARN] Set the environment variable IBM_APPID_TENANT_ID for testing AppID resources, AppID tests will fail if this is not set
[WARN] Set the environment variable IBM_ORG for testing ibm_org  resource Some tests for that resource will fail if this is not set correctly
[WARN] Set the environment variable IBM_SPACE for testing ibm_space  resource Some tests for that resource will fail if this is not set correctly
[WARN] Set the environment variable IBM_ID1 for testing ibm_space resource Some tests for that resource will fail if this is not set correctly
[WARN] Set the environment variable IBM_ID2 for testing ibm_space resource Some tests for that resource will fail if this is not set correctly
[WARN] Set the environment variable IBM_IAMUSER for testing ibm_iam_user_policy resource Some tests for that resource will fail if this is not set correctly
[WARN] Set the environment variable IBM_DATACENTER for testing ibm_container_cluster resource else it is set to default value 'par01'
[WARN] Set the environment variable IBM_MACHINE_TYPE for testing ibm_container_cluster resource else it is set to default value 'u2c.2x4'
[WARN] Set the environment variable IBM_CERT_CRN for testing ibm_container_alb_cert resource else it is set to default value
[WARN] Set the environment variable IBM_UPDATE_CERT_CRN for testing ibm_container_alb_cert resource else it is set to default value
[WARN] Set the environment variable IBM_CONTAINER_REGION for testing ibm_container resources else it is set to default value 'eu-de'
[WARN] Set the environment variable IBM_CIS_INSTANCE with a VALID CIS Instance NAME for testing ibm_cis resources on staging/test
[WARN] Set the environment variable IBM_CIS_DOMAIN_STATIC with the Domain name registered with the CIS instance on test/staging. Domain must be predefined in CIS to avoid CIS billing costs due to domain delete/create
[WARN] Set the environment variable IBM_CIS_DOMAIN_TEST with a VALID Domain name for testing the one time create and delete of a domain in CIS. Note each create/delete will trigger a monthly billing instance. Only to be run in staging/test
[WARN] Set the environment variable IBM_CIS_RESOURCE_GROUP with the resource group for the CIS Instance
[WARN] Set the environment variable IBM_COS_CRN with a VALID COS instance CRN for testing ibm_cos_* resources
[WARN] Set the environment variable IBM_TRUSTED_MACHINE_TYPE for testing ibm_container_cluster resource else it is set to default value 'mb1c.16x64'
[WARN] Set the environment variable IBM_BM_EXTENDED_HW_TESTING to true/false for testing ibm_compute_bare_metal resource else it is set to default value 'false'
[WARN] Set the environment variable IBM_PUBLIC_VLAN_ID for testing ibm_container_cluster resource else it is set to default value '2393319'
[WARN] Set the environment variable IBM_PRIVATE_VLAN_ID for testing ibm_container_cluster resource else it is set to default value '2393321'
[WARN] Set the environment variable IBM_KUBE_VERSION for testing ibm_container_cluster resource else it is set to default value '1.18.14'
[WARN] Set the environment variable IBM_KUBE_UPDATE_VERSION for testing ibm_container_cluster resource else it is set to default value '1.19.6'
[WARN] Set the environment variable IBM_PRIVATE_SUBNET_ID for testing ibm_container_cluster resource else it is set to default value '1636107'
[WARN] Set the environment variable IBM_PUBLIC_SUBNET_ID for testing ibm_container_cluster resource else it is set to default value '1165645'
[WARN] Set the environment variable IBM_SUBNET_ID for testing ibm_container_cluster resource else it is set to default value '1165645'
[INFO] Set the environment variable IBM_IPSEC_DATACENTER for testing ibm_ipsec_vpn resource else it is set to default value 'tok02'
[INFO] Set the environment variable IBM_IPSEC_CUSTOMER_SUBNET_ID for testing ibm_ipsec_vpn resource else it is set to default value '123456'
[INFO] Set the environment variable IBM_IPSEC_CUSTOMER_PEER_IP for testing ibm_ipsec_vpn resource else it is set to default value '192.168.0.1'
[WARN] Set the environment variable IBM_LBAAS_DATACENTER for testing ibm_lbaas resource else it is set to default value 'dal13'
[WARN] Set the environment variable IBM_LBAAS_SUBNETID for testing ibm_lbaas resource else it is set to default value '2144241'
[WARN] Set the environment variable IBM_DEDICATED_HOSTNAME for testing ibm_compute_vm_instance resource else it is set to default value 'terraform-dedicatedhost'
[WARN] Set the environment variable IBM_DEDICATED_HOST_ID for testing ibm_compute_vm_instance resource else it is set to default value '30301'
[WARN] Set the environment variable IBM_WORKER_POOL_ZONE for testing ibm_container_worker_pool_zone_attachment resource else it is set to default value 'ams03'
[WARN] Set the environment variable IBM_WORKER_POOL_ZONE_PRIVATE_VLAN for testing ibm_container_worker_pool_zone_attachment resource else it is set to default value '2538975'
[WARN] Set the environment variable IBM_WORKER_POOL_ZONE_PUBLIC_VLAN for testing ibm_container_worker_pool_zone_attachment resource else it is set to default value '2538967'
[WARN] Set the environment variable IBM_WORKER_POOL_ZONE_UPDATE_PRIVATE_VLAN for testing ibm_container_worker_pool_zone_attachment resource else it is set to default value '2388377'
[WARN] Set the environment variable IBM_WORKER_POOL_ZONE_UPDATE_PUBLIC_VLAN for testing ibm_container_worker_pool_zone_attachment resource else it is set to default value '2388375'
[WARN] Set the environment variable IBM_PLACEMENT_GROUP_NAME for testing ibm_compute_vm_instance resource else it is set to default value 'terraform-group'
[INFO] Set the environment variable SL_REGION for testing ibm_is_region datasource else it is set to default value 'us-south'
[INFO] Set the environment variable SL_ZONE for testing ibm_is_zone datasource else it is set to default value 'us-south-1'
[INFO] Set the environment variable SL_CIDR for testing ibm_is_subnet else it is set to default value '10.240.0.0/24'
[INFO] Set the environment variable SL_ADDRESS_PREFIX_CIDR for testing ibm_is_vpc_address_prefix else it is set to default value '10.120.0.0/24'
[INFO] Set the environment variable IS_IMAGE for testing ibm_is_instance, ibm_is_floating_ip else it is set to default value 'r006-ed3f775f-ad7e-4e37-ae62-7199b4988b00'
[INFO] Set the environment variable IS_WIN_IMAGE for testing ibm_is_instance data source else it is set to default value 'r006-5f9568ae-792e-47e1-a710-5538b2bdfca7'
[INFO] Set the environment variable SL_INSTANCE_PROFILE for testing ibm_is_instance resource else it is set to default value 'cx2-2x4'
[INFO] Set the environment variable SL_INSTANCE_PROFILE_UPDATE for testing ibm_is_instance resource else it is set to default value 'cx2-4x8'
[INFO] Set the environment variable IS_DEDICATED_HOST_NAME for testing ibm_is_instance resource else it is set to default value 'tf-dhost-01'
[INFO] Set the environment variable IS_DEDICATED_HOST_GROUP_ID for testing ibm_is_instance resource else it is set to default value '0717-9104e7b5-77ad-44ad-9eaa-091e6b6efce1'
[INFO] Set the environment variable IS_DEDICATED_HOST_PROFILE for testing ibm_is_instance resource else it is set to default value 'bx2d-host-152x608'
[INFO] Set the environment variable IS_DEDICATED_HOST_GROUP_CLASS for testing ibm_is_instance resource else it is set to default value 'bx2d'
[INFO] Set the environment variable IS_DEDICATED_HOST_GROUP_FAMILY for testing ibm_is_instance resource else it is set to default value 'balanced'
[INFO] Set the environment variable SL_INSTANCE_PROFILE for testing ibm_is_instance resource else it is set to default value 'bx2d-16x64'
[INFO] Set the environment variable IS_VOLUME_PROFILE for testing ibm_is_volume_profile else it is set to default value 'general-purpose'
[INFO] Set the environment variable SL_ROUTE_DESTINATION for testing ibm_is_vpc_route else it is set to default value '192.168.4.0/24'
[INFO] Set the environment variable SL_ROUTE_NEXTHOP for testing ibm_is_vpc_route else it is set to default value '10.0.0.4'
[INFO] Set the environment variable PI_IMAGE for testing ibm_pi_image resource else it is set to default value '7200-03-03'
[INFO] Set the environment variable PI_KEY_NAME for testing ibm_pi_key_name resource else it is set to default value 'terraform-test-power'
[INFO] Set the environment variable PI_NETWORK_NAME for testing ibm_pi_network_name resource else it is set to default value 'terraform-test-power'
[INFO] Set the environment variable PI_VOLUME_NAME for testing ibm_pi_network_name resource else it is set to default value 'terraform-test-power'
[INFO] Set the environment variable PI_CLOUDINSTANCE_ID for testing ibm_pi_image resource else it is set to default value 'd16705bd-7f1a-48c9-9e0e-1c17b71e7331'
[INFO] Set the environment variable PI_PVM_INSTANCE_ID for testing pi_instance_name resource else it is set to default value 'terraform-test-power'
[INFO] Set the environment variable SCHEMATICS_WORKSPACE_ID for testing schematics resources else it is set to default value
[INFO] Set the environment variable SCHEMATICS_TEMPLATE_ID for testing schematics resources else it is set to default value
[INFO] Set the environment variable SCHEMATICS_ACTION_ID for testing schematics resources else it is set to default value
[WARN] Set the environment variable IMAGE_COS_URL with a VALID COS Image SQL URL for testing ibm_is_image resources on staging/test
[WARN] Set the environment variable IMAGE_COS_URL_ENCRYPTED with a VALID COS Image SQL URL for testing ibm_is_image resources on staging/test
[WARN] Set the environment variable IMAGE_OPERATING_SYSTEM with a VALID Operating system for testing ibm_is_image resources on staging/test
[INFO] Set the environment variable IS_IMAGE_NAME for testing data source ibm_is_image else it is set to default value `ibm-ubuntu-18-04-1-minimal-amd64-2`
[INFO] Set the environment variable IS_IMAGE_ENCRYPTED_DATA_KEY for testing resource ibm_is_image else it is set to default value
[INFO] Set the environment variable IS_IMAGE_ENCRYPTION_KEY for testing resource ibm_is_image else it is set to default value
[INFO] Set the environment variable IBM_FUNCTION_NAMESPACE for testing ibm_function_package, ibm_function_action, ibm_function_rule, ibm_function_trigger resource else  tests will fail if this is not set correctly
[INFO] Set the environment variable HPCS_INSTANCE_ID for testing data_source_ibm_kms_key_test else it is set to default value
[INFO] Set the environment variable SECRETS_MANAGER_INSTANCE_ID for testing data_source_ibm_secrets_manager_secrets_test else tests will fail if this is not set correctly
[INFO] Set the environment variable SECRETS_MANAGER_SECRET_TYPE for testing data_source_ibm_secrets_manager_secrets_test, else it is set to default value. For data_source_ibm_secrets_manager_secret_test, tests will fail if this is not set correctly
[WARN] Set the environment variable SECRETS_MANAGER_SECRET_ID for testing data_source_ibm_secrets_manager_secret_test else tests will fail if this is not set correctly
[INFO] Set the environment variable IBM_TG_CROSS_ACCOUNT_ID for testing ibm_tg_connection resource else  tests will fail if this is not set correctly
[INFO] Set the environment variable IBM_TG_CROSS_NETWORK_ID for testing ibm_tg_connection resource else  tests will fail if this is not set correctly
[INFO] Set the environment variable ACCOUNT_TO_BE_IMPORTED for testing import enterprise account resource else  tests will fail if this is not set correctly
[WARN] Set the environment variable IBM_HPCS_ADMIN1 with a VALID HPCS Admin Key1 Path
[WARN] Set the environment variable IBM_HPCS_TOKEN1 with a VALID token for HPCS Admin Key1
[WARN] Set the environment variable IBM_HPCS_ADMIN2 with a VALID HPCS Admin Key2 Path
[WARN] Set the environment variable IBM_HPCS_TOKEN2 with a VALID token for HPCS Admin Key2
=== RUN   TestAccIBMCloudantDataSource_basic
--- PASS: TestAccIBMCloudantDataSource_basic (106.78s)
=== RUN   TestAccIBMCloudant_basic
--- PASS: TestAccIBMCloudant_basic (195.36s)
=== RUN   TestAccIBMCloudant_import
--- PASS: TestAccIBMCloudant_import (104.85s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm 408.466s
...
```

kavya498 commented on ibm/resource_ibm_cloudant.go:
```diff
@@ -0,0 +1,586 @@
+
+	client, err := cloudantv1.NewCloudantV1(&cloudantv1.CloudantV1Options{
+		Authenticator: &core.BearerTokenAuthenticator{
+			BearerToken: token,
```
Please include apikey authentication as well
something similar to https://github.com/IBM-Cloud/terraform-provider-ibm/blob/master/ibm/config.go#L1161

kavya498 commented on website/docs/d/cloudant.html.markdown:
```diff
@@ -0,0 +1,61 @@
+* `location` (Optional, String) The location or the environment in which instance exists.
+* `resource_group_id` (Optional, String) The id of the resource group in which the instance is present. If not provided it takes the default resource group.
+
+## Attribute Reference
```
Please re-arrange argument or attribute reference to alphabetical order..

eiri replied:
Addressed in https://github.com/IBM-Cloud/terraform-provider-ibm/pull/3065/commits/4c6eb75ec54b0fe6fe1b508a6af8ad6aeecbbced

kavya498 commented:
Please resolve all conflicts

eiri commented:
Output from acceptance testing after updates:

```bash
$ make testacc TEST=./ibm TESTARGS='-run TestAccIBMCloudant'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./ibm -v -run TestAccIBMCloudant -timeout 700m
[WARN] Set the environment variable IBM_APPID_TENANT_ID for testing AppID resources, AppID tests will fail if this is not set
[WARN] Set the environment variable IBM_ORG for testing ibm_org  resource Some tests for that resource will fail if this is not set correctly
[WARN] Set the environment variable IBM_SPACE for testing ibm_space  resource Some tests for that resource will fail if this is not set correctly
[WARN] Set the environment variable IBM_ID1 for testing ibm_space resource Some tests for that resource will fail if this is not set correctly
[WARN] Set the environment variable IBM_ID2 for testing ibm_space resource Some tests for that resource will fail if this is not set correctly
[WARN] Set the environment variable IBM_IAMUSER for testing ibm_iam_user_policy resource Some tests for that resource will fail if this is not set correctly
[WARN] Set the environment variable IBM_DATACENTER for testing ibm_container_cluster resource else it is set to default value 'par01'
[WARN] Set the environment variable IBM_MACHINE_TYPE for testing ibm_container_cluster resource else it is set to default value 'u2c.2x4'
[WARN] Set the environment variable IBM_CERT_CRN for testing ibm_container_alb_cert resource else it is set to default value
[WARN] Set the environment variable IBM_UPDATE_CERT_CRN for testing ibm_container_alb_cert resource else it is set to default value
[WARN] Set the environment variable IBM_CONTAINER_REGION for testing ibm_container resources else it is set to default value 'eu-de'
[WARN] Set the environment variable IBM_CIS_INSTANCE with a VALID CIS Instance NAME for testing ibm_cis resources on staging/test
[WARN] Set the environment variable IBM_CIS_DOMAIN_STATIC with the Domain name registered with the CIS instance on test/staging. Domain must be predefined in CIS to avoid CIS billing costs due to domain delete/create
[WARN] Set the environment variable IBM_CIS_DOMAIN_TEST with a VALID Domain name for testing the one time create and delete of a domain in CIS. Note each create/delete will trigger a monthly billing instance. Only to be run in staging/test
[WARN] Set the environment variable IBM_CIS_RESOURCE_GROUP with the resource group for the CIS Instance
[WARN] Set the environment variable IBM_COS_CRN with a VALID COS instance CRN for testing ibm_cos_* resources
[WARN] Set the environment variable IBM_TRUSTED_MACHINE_TYPE for testing ibm_container_cluster resource else it is set to default value 'mb1c.16x64'
[WARN] Set the environment variable IBM_BM_EXTENDED_HW_TESTING to true/false for testing ibm_compute_bare_metal resource else it is set to default value 'false'
[WARN] Set the environment variable IBM_PUBLIC_VLAN_ID for testing ibm_container_cluster resource else it is set to default value '2393319'
[WARN] Set the environment variable IBM_PRIVATE_VLAN_ID for testing ibm_container_cluster resource else it is set to default value '2393321'
[WARN] Set the environment variable IBM_KUBE_VERSION for testing ibm_container_cluster resource else it is set to default value '1.18.14'
[WARN] Set the environment variable IBM_KUBE_UPDATE_VERSION for testing ibm_container_cluster resource else it is set to default value '1.19.6'
[WARN] Set the environment variable IBM_PRIVATE_SUBNET_ID for testing ibm_container_cluster resource else it is set to default value '1636107'
[WARN] Set the environment variable IBM_PUBLIC_SUBNET_ID for testing ibm_container_cluster resource else it is set to default value '1165645'
[WARN] Set the environment variable IBM_SUBNET_ID for testing ibm_container_cluster resource else it is set to default value '1165645'
[INFO] Set the environment variable IBM_IPSEC_DATACENTER for testing ibm_ipsec_vpn resource else it is set to default value 'tok02'
[INFO] Set the environment variable IBM_IPSEC_CUSTOMER_SUBNET_ID for testing ibm_ipsec_vpn resource else it is set to default value '123456'
[INFO] Set the environment variable IBM_IPSEC_CUSTOMER_PEER_IP for testing ibm_ipsec_vpn resource else it is set to default value '192.168.0.1'
[WARN] Set the environment variable IBM_LBAAS_DATACENTER for testing ibm_lbaas resource else it is set to default value 'dal13'
[WARN] Set the environment variable IBM_LBAAS_SUBNETID for testing ibm_lbaas resource else it is set to default value '2144241'
[WARN] Set the environment variable IBM_DEDICATED_HOSTNAME for testing ibm_compute_vm_instance resource else it is set to default value 'terraform-dedicatedhost'
[WARN] Set the environment variable IBM_DEDICATED_HOST_ID for testing ibm_compute_vm_instance resource else it is set to default value '30301'
[WARN] Set the environment variable IBM_WORKER_POOL_ZONE for testing ibm_container_worker_pool_zone_attachment resource else it is set to default value 'ams03'
[WARN] Set the environment variable IBM_WORKER_POOL_ZONE_PRIVATE_VLAN for testing ibm_container_worker_pool_zone_attachment resource else it is set to default value '2538975'
[WARN] Set the environment variable IBM_WORKER_POOL_ZONE_PUBLIC_VLAN for testing ibm_container_worker_pool_zone_attachment resource else it is set to default value '2538967'
[WARN] Set the environment variable IBM_WORKER_POOL_ZONE_UPDATE_PRIVATE_VLAN for testing ibm_container_worker_pool_zone_attachment resource else it is set to default value '2388377'
[WARN] Set the environment variable IBM_WORKER_POOL_ZONE_UPDATE_PUBLIC_VLAN for testing ibm_container_worker_pool_zone_attachment resource else it is set to default value '2388375'
[WARN] Set the environment variable IBM_PLACEMENT_GROUP_NAME for testing ibm_compute_vm_instance resource else it is set to default value 'terraform-group'
[INFO] Set the environment variable SL_REGION for testing ibm_is_region datasource else it is set to default value 'us-south'
[INFO] Set the environment variable SL_ZONE for testing ibm_is_zone datasource else it is set to default value 'us-south-1'
[INFO] Set the environment variable SL_CIDR for testing ibm_is_subnet else it is set to default value '10.240.0.0/24'
[INFO] Set the environment variable SL_ADDRESS_PREFIX_CIDR for testing ibm_is_vpc_address_prefix else it is set to default value '10.120.0.0/24'
[INFO] Set the environment variable IS_IMAGE for testing ibm_is_instance, ibm_is_floating_ip else it is set to default value 'r006-ed3f775f-ad7e-4e37-ae62-7199b4988b00'
[INFO] Set the environment variable IS_WIN_IMAGE for testing ibm_is_instance data source else it is set to default value 'r006-5f9568ae-792e-47e1-a710-5538b2bdfca7'
[INFO] Set the environment variable SL_INSTANCE_PROFILE for testing ibm_is_instance resource else it is set to default value 'cx2-2x4'
[INFO] Set the environment variable SL_INSTANCE_PROFILE_UPDATE for testing ibm_is_instance resource else it is set to default value 'cx2-4x8'
[INFO] Set the environment variable IS_DEDICATED_HOST_NAME for testing ibm_is_instance resource else it is set to default value 'tf-dhost-01'
[INFO] Set the environment variable IS_DEDICATED_HOST_GROUP_ID for testing ibm_is_instance resource else it is set to default value '0717-9104e7b5-77ad-44ad-9eaa-091e6b6efce1'
[INFO] Set the environment variable IS_DEDICATED_HOST_PROFILE for testing ibm_is_instance resource else it is set to default value 'bx2d-host-152x608'
[INFO] Set the environment variable IS_DEDICATED_HOST_GROUP_CLASS for testing ibm_is_instance resource else it is set to default value 'bx2d'
[INFO] Set the environment variable IS_DEDICATED_HOST_GROUP_FAMILY for testing ibm_is_instance resource else it is set to default value 'balanced'
[INFO] Set the environment variable SL_INSTANCE_PROFILE for testing ibm_is_instance resource else it is set to default value 'bx2d-16x64'
[INFO] Set the environment variable IS_VOLUME_PROFILE for testing ibm_is_volume_profile else it is set to default value 'general-purpose'
[INFO] Set the environment variable SL_ROUTE_DESTINATION for testing ibm_is_vpc_route else it is set to default value '192.168.4.0/24'
[INFO] Set the environment variable SL_ROUTE_NEXTHOP for testing ibm_is_vpc_route else it is set to default value '10.0.0.4'
[INFO] Set the environment variable PI_IMAGE for testing ibm_pi_image resource else it is set to default value '7200-03-03'
[INFO] Set the environment variable PI_KEY_NAME for testing ibm_pi_key_name resource else it is set to default value 'terraform-test-power'
[INFO] Set the environment variable PI_NETWORK_NAME for testing ibm_pi_network_name resource else it is set to default value 'terraform-test-power'
[INFO] Set the environment variable PI_VOLUME_NAME for testing ibm_pi_network_name resource else it is set to default value 'terraform-test-power'
[INFO] Set the environment variable PI_CLOUDINSTANCE_ID for testing ibm_pi_image resource else it is set to default value 'd16705bd-7f1a-48c9-9e0e-1c17b71e7331'
[INFO] Set the environment variable PI_PVM_INSTANCE_ID for testing pi_instance_name resource else it is set to default value 'terraform-test-power'
[INFO] Set the environment variable SCHEMATICS_WORKSPACE_ID for testing schematics resources else it is set to default value
[INFO] Set the environment variable SCHEMATICS_TEMPLATE_ID for testing schematics resources else it is set to default value
[INFO] Set the environment variable SCHEMATICS_ACTION_ID for testing schematics resources else it is set to default value
[WARN] Set the environment variable IMAGE_COS_URL with a VALID COS Image SQL URL for testing ibm_is_image resources on staging/test
[WARN] Set the environment variable IMAGE_COS_URL_ENCRYPTED with a VALID COS Image SQL URL for testing ibm_is_image resources on staging/test
[WARN] Set the environment variable IMAGE_OPERATING_SYSTEM with a VALID Operating system for testing ibm_is_image resources on staging/test
[INFO] Set the environment variable IS_IMAGE_NAME for testing data source ibm_is_image else it is set to default value `ibm-ubuntu-18-04-1-minimal-amd64-2`
[INFO] Set the environment variable IS_IMAGE_ENCRYPTED_DATA_KEY for testing resource ibm_is_image else it is set to default value
[INFO] Set the environment variable IS_IMAGE_ENCRYPTION_KEY for testing resource ibm_is_image else it is set to default value
[INFO] Set the environment variable IBM_FUNCTION_NAMESPACE for testing ibm_function_package, ibm_function_action, ibm_function_rule, ibm_function_trigger resource else  tests will fail if this is not set correctly
[INFO] Set the environment variable HPCS_INSTANCE_ID for testing data_source_ibm_kms_key_test else it is set to default value
[INFO] Set the environment variable SECRETS_MANAGER_INSTANCE_ID for testing data_source_ibm_secrets_manager_secrets_test else tests will fail if this is not set correctly
[INFO] Set the environment variable SECRETS_MANAGER_SECRET_TYPE for testing data_source_ibm_secrets_manager_secrets_test, else it is set to default value. For data_source_ibm_secrets_manager_secret_test, tests will fail if this is not set correctly
[WARN] Set the environment variable SECRETS_MANAGER_SECRET_ID for testing data_source_ibm_secrets_manager_secret_test else tests will fail if this is not set correctly
[INFO] Set the environment variable IBM_TG_CROSS_ACCOUNT_ID for testing ibm_tg_connection resource else  tests will fail if this is not set correctly
[INFO] Set the environment variable IBM_TG_CROSS_NETWORK_ID for testing ibm_tg_connection resource else  tests will fail if this is not set correctly
[INFO] Set the environment variable ACCOUNT_TO_BE_IMPORTED for testing import enterprise account resource else  tests will fail if this is not set correctly
[WARN] Set the environment variable IBM_HPCS_ADMIN1 with a VALID HPCS Admin Key1 Path
[WARN] Set the environment variable IBM_HPCS_TOKEN1 with a VALID token for HPCS Admin Key1
[WARN] Set the environment variable IBM_HPCS_ADMIN2 with a VALID HPCS Admin Key2 Path
[WARN] Set the environment variable IBM_HPCS_TOKEN2 with a VALID token for HPCS Admin Key2
[INFO] Set the environment variable SCC_SI_ACCOUNT for testing SCC SI resources resource else  tests will fail if this is not set correctly
[INFO] Set the environment variable IBM_CLOUD_SHELL_ACCOUNT_ID for ibm-cloud-shell resource or datasource else tests will fail if this is not set correctly
=== RUN   TestAccIBMCloudantDataSource_basic
--- PASS: TestAccIBMCloudantDataSource_basic (99.49s)
=== RUN   TestAccIBMCloudant_basic
--- PASS: TestAccIBMCloudant_basic (124.95s)
=== RUN   TestAccIBMCloudant_import
--- PASS: TestAccIBMCloudant_import (114.92s)
PASS
ok  	github.com/IBM-Cloud/terraform-provider-ibm/ibm	341.574s
```

geethasathyamurthy commented on website/docs/d/cloudant.html.markdown:
```diff
@@ -0,0 +1,61 @@
+
+# ibm_cloudant
+
+Provides a read-only data source for an existing IBM Cloud Cloudant service. You can then reference the fields of the data source in other resources within the same configuration using interpolation syntax.
```
Could add a reference link sentence like: For more information, about how to use this service, see[Cloudant documentation](https://cloud.ibm.com/docs/Cloudant?topic=Cloudant-getting-started-with-cloudant).

geethasathyamurthy commented on website/docs/d/cloudant.html.markdown:
```diff
@@ -0,0 +1,61 @@
+
+Provides a read-only data source for an existing IBM Cloud Cloudant service. You can then reference the fields of the data source in other resources within the same configuration using interpolation syntax.
+
+## Example Usage
```
U in Usage should be **usage**

geethasathyamurthy commented on website/docs/d/cloudant.html.markdown:
````diff
@@ -0,0 +1,61 @@
+
+## Example Usage
+
+```hcl
````
code tag to contain **terraform** instead of **hcl**

geethasathyamurthy commented on website/docs/d/cloudant.html.markdown:
````diff
@@ -0,0 +1,61 @@
+ }
+```
+
+## Argument Reference
````
R in Reference should be **reference**

geethasathyamurthy commented on website/docs/d/cloudant.html.markdown:
```diff
@@ -0,0 +1,61 @@
+The following arguments are supported:
+
+* `id` (Optional, String) The unique identifier of the Cloudant resource.
+* `location` (Optional, String) The location or the environment in which instance exists.
```
The location or the environment in which instance exists.  should be as
The location or the environment in which the instance exists.

geethasathyamurthy commented on website/docs/d/cloudant.html.markdown:
```diff
@@ -0,0 +1,61 @@
+
+In addition to all arguments above, the following attributes are exported:
+
+* `capacity` (Number) A number of blocks of throughput units. For more details please read about [`blocks`](https://cloud.ibm.com/apidocs/cloudant#putcapacitythroughputconfiguration) parameter.
```
A number of blocks of throughput units. For more details please read about [`blocks`](https://cloud.ibm.com/apidocs/cloudant#putcapacitythroughputconfiguration) parameter.  can be as

 A number of blocks of throughput units. For more information, about throughput capacity, see [blocks parameter](https://cloud.ibm.com/apidocs/cloudant#putcapacitythroughputconfiguration). parameter.

geethasathyamurthy commented on website/docs/d/cloudant.html.markdown:
```diff
@@ -0,0 +1,61 @@
+In addition to all arguments above, the following attributes are exported:
+
+* `capacity` (Number) A number of blocks of throughput units. For more details please read about [`blocks`](https://cloud.ibm.com/apidocs/cloudant#putcapacitythroughputconfiguration) parameter.
+* `cors_config` (List of Object) Configuration for CORS. (see [below for nested attributes](#nestedatt--cors_config))
```
delete (see [below for nested attributes](#nestedatt--cors_config)) and replace as 
  Nested scheme for `cors_config`:

geethasathyamurthy commented on website/docs/d/cloudant.html.markdown:
```diff
@@ -0,0 +1,61 @@
+
+* `capacity` (Number) A number of blocks of throughput units. For more details please read about [`blocks`](https://cloud.ibm.com/apidocs/cloudant#putcapacitythroughputconfiguration) parameter.
+* `cors_config` (List of Object) Configuration for CORS. (see [below for nested attributes](#nestedatt--cors_config))
+    * `allow_credentials` (Boolean) - Boolean value to allow authentication credentials. If set to true, browser requests must be done by using withCredentials = true.
```
The boolean value to allow authentication credentials. If set to **true** ......
This statement is unclear, please add if set to **true**, what happens

geethasathyamurthy commented on website/docs/d/cloudant.html.markdown:
```diff
@@ -0,0 +1,61 @@
+* `cors_config` (List of Object) Configuration for CORS. (see [below for nested attributes](#nestedatt--cors_config))
+    * `allow_credentials` (Boolean) - Boolean value to allow authentication credentials. If set to true, browser requests must be done by using withCredentials = true.
+    * `origins` (List of String) - Contains the list of allowed origin domains with the full URL including the protocol. Subdomains count as separate domains, so all subdomains used have to be listed.
+* `crn` (String) CRN of resource instance.
```
should be: The CRN of the resource instance.

geethasathyamurthy commented on website/docs/d/cloudant.html.markdown:
```diff
@@ -0,0 +1,61 @@
+    * `allow_credentials` (Boolean) - Boolean value to allow authentication credentials. If set to true, browser requests must be done by using withCredentials = true.
+    * `origins` (List of String) - Contains the list of allowed origin domains with the full URL including the protocol. Subdomains count as separate domains, so all subdomains used have to be listed.
+* `crn` (String) CRN of resource instance.
+* `enable_cors` (Boolean) Boolean value to turn CORS on and off.
```
Could be as:
 The boolean value to turn CORS on or off.  The supported values are **on** and **off**.

geethasathyamurthy commented on website/docs/d/cloudant.html.markdown:
```diff
@@ -0,0 +1,61 @@
+* `extensions` (Map of String) The extended metadata as a map associated with the resource instance.
+* `features` (List of String) List of enabled optional features.
+* `features_flags` (List of String) List of feature flags.
+* `guid` (String) Guid of resource instance.
```
Guid of resource instance. could be as
 `GUID` of the resource instance.

geethasathyamurthy commented on website/docs/d/cloudant.html.markdown:
```diff
@@ -0,0 +1,61 @@
+* `features` (List of String) List of enabled optional features.
+* `features_flags` (List of String) List of feature flags.
+* `guid` (String) Guid of resource instance.
+* `include_data_events` (Boolean) Include `data` event types in events sent to IBM Cloud Activity Tracker with LogDNA for the IBM Cloudant instance. By default emitted events are only of `management` type.
```
are only of `management` type.  could be as
are only of the `management` type.

geethasathyamurthy commented on website/docs/d/cloudant.html.markdown:
```diff
@@ -0,0 +1,61 @@
+* `include_data_events` (Boolean) Include `data` event types in events sent to IBM Cloud Activity Tracker with LogDNA for the IBM Cloudant instance. By default emitted events are only of `management` type.
+* `plan` (String) The plan type of the instance.
+* `resource_controller_url` (String) The URL of the IBM Cloud dashboard that can be used to explore and view details about the resource.
+* `resource_crn` (String) The crn of the resource.
```
The crn of the resource.  Could be as
The CRN of the resource.

geethasathyamurthy commented on website/docs/d/cloudant.html.markdown:
```diff
@@ -0,0 +1,61 @@
+* `service` (String) The service type of the instance.
+* `status` (String) The resource instance status.
+* `throughput` (Map of Number) Schema for detailed information about throughput capacity with breakdown by specific throughput requests classes.
+* `version` (String) Vendor version.
```
The Vendor version.

geethasathyamurthy commented on website/docs/r/cloudant.html.markdown:
```diff
@@ -0,0 +1,131 @@
+# ibm_cloudant
+
+Provides a resource for IBM Cloudant. This allows an IBM Cloudant service instance to be created, updated, or deleted.
+For more information, about how to use this service, see the official [Cloudant documentation](https://cloud.ibm.com/docs/Cloudant?topic=Cloudant-getting-started-with-cloudant).
```
For more information, about how to use this service, see the official [Cloudant documentation](https://cloud.ibm.com/docs/Cloudant?topic=Cloudant-getting-started-with-cloudant). 
as
For more information, about Cloudant, see [Getting started with IBM Cloudant](https://cloud.ibm.com/docs/Cloudant?topic=Cloudant-getting-started-with-cloudant).

geethasathyamurthy commented on website/docs/r/cloudant.html.markdown:
```diff
@@ -0,0 +1,131 @@
+Provides a resource for IBM Cloudant. This allows an IBM Cloudant service instance to be created, updated, or deleted.
+For more information, about how to use this service, see the official [Cloudant documentation](https://cloud.ibm.com/docs/Cloudant?topic=Cloudant-getting-started-with-cloudant).
+
+## Example Usage
```
U in Usage should be **usage**

geethasathyamurthy commented on website/docs/r/cloudant.html.markdown:
```diff
@@ -0,0 +1,131 @@
+* `delete` - (Default 10 minutes) The update of the IBM Cloudant instance is considered failed if no response received.
+* `update` - (Default 10 minutes) The deletion of the IBM Cloudant instance is considered failed if no response received.
+
+## Argument Reference
```
R in Reference to be as **reference**

geethasathyamurthy commented on website/docs/r/cloudant.html.markdown:
```diff
@@ -0,0 +1,131 @@
+  plan     = "standard"
+
+  legacy_credentials  = true
+  include_data_events = false
```
**hcl** in code tag should be **terraform**

geethasathyamurthy commented on website/docs/r/cloudant.html.markdown:
```diff
@@ -0,0 +1,131 @@
+
+The following arguments are supported:
+
+* `capacity` - (Optional, Number) A number of blocks of throughput units. For more details please read about [`blocks`](https://cloud.ibm.com/apidocs/cloudant#putcapacitythroughputconfiguration) parameter.
```
Could be replaced as:
A number of blocks of throughput units. For more information, about throughput capacity, see [blocks](https://cloud.ibm.com/apidocs/cloudant#putcapacitythroughputconfiguratio) parameter.

geethasathyamurthy commented on website/docs/r/cloudant.html.markdown:
```diff
@@ -0,0 +1,131 @@
+
+## Argument Reference
+
+The following arguments are supported:
```
The following arguments are supported: should be as

Review the argument reference that you can specify for your resource.

geethasathyamurthy commented on website/docs/r/cloudant.html.markdown:
```diff
@@ -0,0 +1,131 @@
+The following arguments are supported:
+
+* `capacity` - (Optional, Number) A number of blocks of throughput units. For more details please read about [`blocks`](https://cloud.ibm.com/apidocs/cloudant#putcapacitythroughputconfiguration) parameter.
+  * Constraints: The default value is `1`. Capacity modification is not supported for `lite` plan.
```
If Constraints is a nested argument for capacity, then you should provide as
Nested scheme for `capacity`:
Else

Constraints parameter bullet need to be aligned.

geethasathyamurthy commented on website/docs/r/cloudant.html.markdown:
```diff
@@ -0,0 +1,131 @@
+
+* `capacity` - (Optional, Number) A number of blocks of throughput units. For more details please read about [`blocks`](https://cloud.ibm.com/apidocs/cloudant#putcapacitythroughputconfiguration) parameter.
+  * Constraints: The default value is `1`. Capacity modification is not supported for `lite` plan.
+* `cors_config` - (Optional, Block List) Configuration for CORS. (see [below for nested arguments](#nestedblock--cors_config))
```
Replace this statement (see [below for nested arguments](#nestedblock--cors_config)). (see [below for nested arguments](#nestedblock--cors_config))

Nested scheme for `cors_config`:

geethasathyamurthy commented on website/docs/r/cloudant.html.markdown:
```diff
@@ -0,0 +1,131 @@
+* `capacity` - (Optional, Number) A number of blocks of throughput units. For more details please read about [`blocks`](https://cloud.ibm.com/apidocs/cloudant#putcapacitythroughputconfiguration) parameter.
+  * Constraints: The default value is `1`. Capacity modification is not supported for `lite` plan.
+* `cors_config` - (Optional, Block List) Configuration for CORS. (see [below for nested arguments](#nestedblock--cors_config))
+  * Constraints: The minimum length is `1` item. Can conflict with `enable_cors` if it is set to `false`. In this case the CORS customization is not allowed.
```
The minimum length is `1` item. Can conflict with `enable_cors` if it is set to `false`. In this case the CORS customization is not allowed.  can be written as

The minimum length is **1** item. This argument can conflict with `enable_cors`. If the argument is set to **false**, the CORS customization is not allowed.

geethasathyamurthy commented on website/docs/r/cloudant.html.markdown:
```diff
@@ -0,0 +1,131 @@
+  * Constraints: The default value is `1`. Capacity modification is not supported for `lite` plan.
+* `cors_config` - (Optional, Block List) Configuration for CORS. (see [below for nested arguments](#nestedblock--cors_config))
+  * Constraints: The minimum length is `1` item. Can conflict with `enable_cors` if it is set to `false`. In this case the CORS customization is not allowed.
+  * `allow_credentials` - (Optional, Boolean) Boolean value to allow authentication credentials. If set to true, browser requests must be done by using withCredentials = true.
```
If set to true, browser requests must be done by using withCredentials = true.  could be as

If set to **true**, browser requests must be done by using `withCredentials = true`.

geethasathyamurthy commented on website/docs/r/cloudant.html.markdown:
```diff
@@ -0,0 +1,131 @@
+* `cors_config` - (Optional, Block List) Configuration for CORS. (see [below for nested arguments](#nestedblock--cors_config))
+  * Constraints: The minimum length is `1` item. Can conflict with `enable_cors` if it is set to `false`. In this case the CORS customization is not allowed.
+  * `allow_credentials` - (Optional, Boolean) Boolean value to allow authentication credentials. If set to true, browser requests must be done by using withCredentials = true.
+    * Constraints: The default value is `true`.
```
There are two Constraints parameters for `cors_config`

geethasathyamurthy commented on website/docs/r/cloudant.html.markdown:
```diff
@@ -0,0 +1,131 @@
+  * `allow_credentials` - (Optional, Boolean) Boolean value to allow authentication credentials. If set to true, browser requests must be done by using withCredentials = true.
+    * Constraints: The default value is `true`.
+  * `origins` - (Required, List of String) An array of strings that contain allowed origin domains. You have to specify the full URL including the protocol. It is recommended that only the HTTPS protocol is used. Subdomains count as separate domains, so you have to specify all subdomains used.
+* `enable_cors` - (Optional, Boolean) Boolean value to turn CORS on and off.
```
The boolean value to turn CORS on or off.  The supported values are **on** and **off**.

geethasathyamurthy commented on website/docs/r/cloudant.html.markdown:
```diff
@@ -0,0 +1,131 @@
+    * Constraints: The default value is `true`.
+  * `origins` - (Required, List of String) An array of strings that contain allowed origin domains. You have to specify the full URL including the protocol. It is recommended that only the HTTPS protocol is used. Subdomains count as separate domains, so you have to specify all subdomains used.
+* `enable_cors` - (Optional, Boolean) Boolean value to turn CORS on and off.
+  * Constraints: The default value is `true`. If it is set to `false`, then customizing `cors_config` is not allowed.
```
Constraints should have a leading sentence as 
Nested scheme for `enable_cors`:
* Constraints: The default value is **true**. If it is set to **false**, then customizing `cors_config` is not allowed.

geethasathyamurthy commented on website/docs/r/cloudant.html.markdown:
```diff
@@ -0,0 +1,131 @@
+  * Constraints: The default value is `true`. If it is set to `false`, then customizing `cors_config` is not allowed.
+* `environment_crn` - (Optional, Forces new resource, String) CRN of the IBM Cloudant Dedicated Hardware plan instance.
+* `id` - (Optional, String) The unique identifier of the new Cloudant resource.
+* `include_data_events` - (Optional, Boolean) Include `data` event types in events sent to IBM Cloud Activity Tracker with LogDNA for the IBM Cloudant instance. By default emitted events are only of `management` type.
```
By default emitted events are only of `management` type. could be as

By default emitted events are only of the `management` type.

geethasathyamurthy commented on website/docs/r/cloudant.html.markdown:
```diff
@@ -0,0 +1,131 @@
+* `environment_crn` - (Optional, Forces new resource, String) CRN of the IBM Cloudant Dedicated Hardware plan instance.
+* `id` - (Optional, String) The unique identifier of the new Cloudant resource.
+* `include_data_events` - (Optional, Boolean) Include `data` event types in events sent to IBM Cloud Activity Tracker with LogDNA for the IBM Cloudant instance. By default emitted events are only of `management` type.
+  * Constraints: The default value is `false`.
```
Constraints should have a leading sentence as 
Nested scheme for `enable_cors`:
* Constraints: The default value is  **false**.

geethasathyamurthy commented on website/docs/r/cloudant.html.markdown:
```diff
@@ -0,0 +1,131 @@
+* `include_data_events` - (Optional, Boolean) Include `data` event types in events sent to IBM Cloud Activity Tracker with LogDNA for the IBM Cloudant instance. By default emitted events are only of `management` type.
+  * Constraints: The default value is `false`.
+* `legacy_credentials` - (Optional, Forces new resource, Boolean) Use both legacy credentials and IAM for authentication.
+  * Constraints: The default value is `false`.
```
Constraints should have a leading sentence as 
Nested scheme for `enable_cors`:
* Constraints: The default value is  **false**.

geethasathyamurthy commented on website/docs/r/cloudant.html.markdown:
```diff
@@ -0,0 +1,131 @@
+* `name` - (Required, String) A name for the resource instance.
+* `parameters` - (Optional, Forces new resource, Map) Arbitrary parameters to pass. Must be a JSON object.
+* `plan` - (Required, String) The plan type of the service.
+* `resource_group_id` - (Optional, Forces new resource, String) The resource group id.
```
The resource group id.  should be as 

The resource group ID.

geethasathyamurthy commented on website/docs/r/cloudant.html.markdown:
```diff
@@ -0,0 +1,131 @@
+* `service_endpoints` - (Optional, String) Types of the service endpoints. Possible values are 'public', 'private', 'public-and-private'.
+* `tags` - (Optional, Set of String) Tags associated with the instance.
+
+## Attribute Reference
```
R in Reference to be as **reference**

geethasathyamurthy commented on website/docs/r/cloudant.html.markdown:
```diff
@@ -0,0 +1,131 @@
+
+## Attribute Reference
+
+In addition to all arguments above, you can access the following attribute references after your resource is created.:
```
Need to remove **:** after **created.**

geethasathyamurthy commented on website/docs/r/cloudant.html.markdown:
```diff
@@ -0,0 +1,131 @@
+* `allow_cleanup` - (Boolean) A boolean that dictates if the resource instance should be deleted (cleaned up) during the processing of a region instance delete call.
+* `created_at` - (String) The date when the instance was created.
+* `created_by` - (String) The subject who created the instance.
+* `crn` - (String) CRN of resource instance.
```
CRN of resource instance.  could be as

 CRN of the resource instance.

geethasathyamurthy commented on website/docs/r/cloudant.html.markdown:
```diff
@@ -0,0 +1,131 @@
+* `created_at` - (String) The date when the instance was created.
+* `created_by` - (String) The subject who created the instance.
+* `crn` - (String) CRN of resource instance.
+* `dashboard_url` - (String) Dashboard URL to access resource.
```
Dashboard URL to access resource. could be as

The dashboard URL to access resources.

geethasathyamurthy commented on website/docs/r/cloudant.html.markdown:
```diff
@@ -0,0 +1,131 @@
+* `deleted_at` - (String) The date when the instance was deleted.
+* `deleted_by` - (String) The subject who deleted the instance.
+* `extensions` - (Map) The extended metadata as a map associated with the resource instance.
+* `guid` - (String) Guid of resource instance.
```
Guid of resource instance. could be as

The GUID of the resource instance.

geethasathyamurthy commented on website/docs/r/cloudant.html.markdown:
```diff
@@ -0,0 +1,131 @@
+* `resource_aliases_url` - (String) The relative path to the resource aliases for the instance.
+* `resource_bindings_url` - (String) The relative path to the resource bindings for the instance.
+* `resource_controller_url` - (String) The URL of the IBM Cloud dashboard that can be used to explore and view details about the resource.
+* `resource_crn` - (String) The crn of the resource.
```
The crn of the resource. could be as

The CRN of the resource.

geethasathyamurthy commented on website/docs/r/cloudant.html.markdown:
```diff
@@ -0,0 +1,131 @@
+* `scheduled_reclaim_by` - (String) The subject who initiated the instance reclamation.
+* `service` - (String) The service type of the instance.
+* `state` - (String) The current state of the instance.
+* `status` - (String) Status of resource instance.
```
Status of resource instance.  could be as

Status of the resource instance.

geethasathyamurthy commented on website/docs/r/cloudant.html.markdown:
```diff
@@ -0,0 +1,131 @@
+* `service` - (String) The service type of the instance.
+* `state` - (String) The current state of the instance.
+* `status` - (String) Status of resource instance.
+* `sub_type` - (String) The sub-type of instance, e.g. cfaas .
```
The sub-type of instance, e.g. cfaas .  could be as

The sub-type of an instance. For example **cfaas**.

geethasathyamurthy commented on website/docs/r/cloudant.html.markdown:
```diff
@@ -0,0 +1,131 @@
+* `sub_type` - (String) The sub-type of instance, e.g. cfaas .
+* `target_crn` - (String) The full deployment CRN as defined in the global catalog.
+* `throughput` - (Map of Number) Schema for detailed information about throughput capacity with breakdown by specific throughput requests classes.
+* `type` - (String) The type of the instance, e.g. service_instance.
```
The type of the instance, e.g. service_instance.  as

The type of the instance. For example, **service_instance**.

geethasathyamurthy commented on website/docs/r/cloudant.html.markdown:
````diff
@@ -0,0 +1,131 @@
+
+### Syntax
+
+```hcl-terraform
````
remove hcl-terraform

geethasathyamurthy commented on website/docs/r/cloudant.html.markdown:
````diff
@@ -0,0 +1,131 @@
+```
+
+### Example
+```hcl
````
remove **hcl** in the code tag.

geethasathyamurthy reviewed (commented):
Reviewed docs update in /website/docs, request to update the review comments.

eiri commented:
@geethasathyamurthy All comments addressed in https://github.com/IBM-Cloud/terraform-provider-ibm/pull/3065/commits/20c1602fe4084a318b58d7f427c7905b9b024ca1

I removed most of the "Constraints" stanzas to avoid confusion and moved default values into the specs, let me know if this is acceptable.
